### ISSUE TITLE
[Enhancement] Word counting with UAX #29 word boundaries

### ISSUE THREAD

**Issue by omasanori** [Zettlr/Zettlr#2363]
### Description

Writing notes in Japanese with Zettlr works fine but word counting is far from a satisfactory behavior. In Japanese, you will not insert white spaces between words. Thus, Zettlr counts the whole paragraph as *a word.* Precise Japanese word counting is hard, but the [Unicode Standard Annex #29 defines a language-agnostic algorithm to determine word boundaries.](https://unicode.org/reports/tr29/#Word_Boundaries) This algorithm will not harm users that is already fine with the current behavior, and help users frustrated by the current behavior more or less.

### Proposed Changes

Implementing a word counting method based on the UAX 29's default word boundaries algorithm. Making that algorithm default or not depends on the runtime performance of it.

### Caveats

The UAX 29's default word boundaries algorithm is more complex than white space-based one. On the other hand, the algorithm is not optimal and [CLDR](http://cldr.unicode.org/)-based language-specific word boundaries algorithm will produce better results.

It is excellent to me if the CLDR-based method is implemented, but language-agnostic one is better than the status quo.

### Do you Wish to Attempt Implementing this Yourself?

No

### Zettlr Version

Stable (most recent version)

### Your Platform

- [X] Windows
- [ ] macOS
- [ ] Linux

### Operating System Version

Windows 11 (Beta)

**Comment by nathanlesage** [comment]
I did not know about this algorithm but I'll sure as hell have a look and implement it. I did not know that Japanese word boundaries didn't work!

**Comment by omasanori** [comment]
Of course! With the UAX 29 default word boundaries algorithm (calculated by https://github.com/eddieantonio/unicode-default-word-boundary):

- **我輩は猫である。** becomes **我/輩/は/猫/で/あ/る/。** (8 words including punctuation)
- **ゼットルは知識をつなぐ。** becomes **ゼットル/は/知/識/を/つ/な/ぐ/。** (9 words including punctuation)
- **Zettlrはゼットルと読む。** becomes **Zettlr/は/ゼットル/と/読/む/。** (7 words including punctuation)

The specification has [the official test suite](http://www.unicode.org/Public/UCD/latest/ucd/auxiliary/WordBreakTest.txt), too.

**Comment by omasanori** [comment]
Well, Zettlr is an Electron-based app, so it uses V8 as underlying JS engine, right? If so, passing `lang` in the YAML frontmatter or, as a fallback, the UI language to `Intl.Segmenter` will give us language-specific word segmentation for free! 😄

**Comment by omasanori** [comment]
```javascript
> let segmenter = new Intl.Segmenter('ja', {granularity: "word"});
undefined
> function segment(str) {
... for (let {segment, index, isWordLike} of segmenter.segment(str)) {
..... console.log("segment at code units [%d, %d): «%s»%s",
.....   index, index + segment.length,
.....   segment,
.....   isWordLike ? " (word-like)" : ""
.....   );
... }
... }
undefined
> segment("我輩は猫である。");
segment at code units [0, 2): «我輩» (word-like)
segment at code units [2, 3): «は» (word-like)
segment at code units [3, 4): «猫» (word-like)
segment at code units [4, 5): «で» (word-like)
segment at code units [5, 7): «ある» (word-like)
segment at code units [7, 8): «。»
undefined
> segment("ゼットルは知識をつなぐ。");
segment at code units [0, 4): «ゼットル» (word-like)
segment at code units [4, 5): «は» (word-like)
segment at code units [5, 7): «知識» (word-like)
segment at code units [7, 8): «を» (word-like)
segment at code units [8, 11): «つなぐ» (word-like)
segment at code units [11, 12): «。»
undefined
> segment("Zettlrはゼットルと読む。");
segment at code units [0, 6): «Zettlr» (word-like)
segment at code units [6, 7): «は» (word-like)
segment at code units [7, 11): «ゼットル» (word-like)
segment at code units [11, 12): «と» (word-like)
segment at code units [12, 14): «読む» (word-like)
segment at code units [14, 15): «。»
undefined
>

**Comment by nathanlesage** [comment]
Uh, I haven't heard of the Intl.Segmenter o.O That looks quite nice!

> Well, Zettlr is an Electron-based app, so it uses V8 as underlying JS engine, right?

Yess! I mean, in the end it would be much better to use anything native rather than doing it manually. However, the word counter in Zettlr does do a little bit of preprocessing, namely: Remove any Markdown elements (since `##` would otherwise be counted as a word). Question is whether the Segmenter would solve that problem as well.

**Comment by omasanori** [comment]
> However, the word counter in Zettlr does do a little bit of preprocessing, namely: Remove any Markdown elements (since ## would otherwise be counted as a word). Question is whether the Segmenter would solve that problem as well.

If I understand correctly, we can use the `isWordLike` parameter in the example above to skip punctuations and other non-letter symbols.

**Comment by omasanori** [comment]
OK, I checked https://github.com/Zettlr/Zettlr/blob/develop/source/common/util/count-words.js and I think that at least comment removal will still be needed to do by Zettlr.

**Comment by omasanori** [comment]
I have no Electron development environment, but the following quick-and-dirty modified version of `countWords` passes all tests on Node.js v16.8.0.

```javascript
function countWords(words, countChars = false, locale = 'en') {
  if (!words || typeof words !== 'string') return 0

  let tmp = words.split('\n')
  if (tmp[0] === '---') {
    tmp.shift()
    let end = 0
    for (let i = 0; i < tmp.length; i++) {
      if ([ '...', '---' ].includes(tmp[i])) {
        end = i
        break
      }
    }

    if (end > 0) words = tmp.slice(end + 1).join('\n')
  }

  words = words.replace(/<!--.*?-->/gms, '')

  let segmenter = new Intl.Segmenter(locale, {granularity: "word"})

  let segments = []
  for (let {segment, index, isWordLike} of segmenter.segment(words)) {
    if (isWordLike) segments.push(segment)
  }

  words = segments

  if (countChars) words = [...words.join('')]

  return words.length
}

const assert = require('assert')

const countWordsTesters = [
  { 'input': 'Lorem\n\n# Ipsum', 'expected': 2 },
  { 'input': 'Lorem\n\n# Ipsum Dolor', 'expected': 3 },
  { 'input': '\n\n', 'expected': 0 },
  { 'input': '* one\n* two\n* three', 'expected': 3 },
  { 'input': '#', 'expected': 0 },
  { 'input': '---\ntitle: "Some title"\nkeywords:\n  - one\n  - two\n  - three\n...\n\n# Heading\n\nLorem Ipsum dolor, sit amet', 'expected': 6 },
  { 'input': '我輩は猫である。', 'locale': 'ja', 'expected': 5 },
  { 'input': 'ゼットルは知識をつなぐ。', 'locale': 'ja', 'expected': 5 },
  { 'input': 'Zettlrはゼットルと読む。', 'locale': 'ja', 'expected': 5 }
]

for (let test of countWordsTesters) {
  assert.strictEqual(countWords(test.input), test.expected)
}
```

**Comment by nathanlesage** [comment]
That's nice! We might even be able to use the frontmatter extractor (if the runtime speed is still low) so that we don't need to manually remove the YAML frontmatter as it is being done now: https://github.com/Zettlr/Zettlr/blob/develop/source/common/util/extract-yaml-frontmatter.ts
